File: src/main/java/me/andre111/d20server/model/entity/ChatEntry.java
package me.andre111.d20server.model.entity;

import java.time.Instant;

import me.andre111.d20server.message.ExcludeFromMessage;

@SuppressWarnings("unused")
public final class ChatEntry {
	private String text;
	
	@ExcludeFromMessage
	private long[] recipents;
	@ExcludeFromMessage
	private boolean includeGMs;
	
	@ExcludeFromMessage
	private long source;
	
	private long time;
	
	public ChatEntry(String text, long source) {
		this.text = text;
		this.source = source;
		this.time = Instant.now().getEpochSecond();
	}
	public ChatEntry(String text, long source, boolean includeGMs, long...recipents) {
		this.text = text;
		this.recipents = recipents;
		this.includeGMs = includeGMs;
		this.source = source;
		this.time = Instant.now().getEpochSecond();
	}
	
	public long[] getRecipents() {
		return recipents;
	}
	
	public boolean doIncludeGMs() {
		return includeGMs;
	}
}

File: core/common/message/chat/chat-entry.js
import { ID } from '../../entity/id.js';
import { registerType } from '../../util/datautil.js';

export class ChatEntry {
    id;
    text;
    time;

    replaceParent;

    recipents;
    includeGMs;
    source;

    rolls;
    triggeredContent;

    constructor(text, source, includeGMs, recipents) {
        if(text) {
            this.id = ID.next();
            this.text = text;
            this.resetTime();
            
            this.recipents = recipents;
            this.includeGMs = includeGMs || false;
            this.source = source;
        }
    }

    // getters
    getID() {
        return this.id;
    }

    getText() {
        return this.text;
    }

    getTime() {
        return this.time;
    }

    getReplaceParent() {
        return this.replaceParent;
    }

    getRecipents() {
        return this.recipents;
    }

    doIncludeGMs() {
        return this.includeGMs;
    }

    getSource() {
        return this.source;
    }

	// special set methods for rolls and triggered data
    getRolls() {
        return this.rolls;
    }

    setRolls(rolls) {
        this.rolls = rolls;
    }

    getTriggeredContent() {
        return this.triggeredContent;
    }

    setTriggeredContent(triggeredContent) {
        for(const tc of triggeredContent) {
			// make entry an actual replace order, instead of normal entry
            tc.parent = this.getID();
            tc.entry.replaceParent = this.getID();
			// transfer important values from parent
            tc.entry.recipents = this.recipents;
            tc.entry.includeGMs = this.includeGMs;
        }

        this.triggeredContent = triggeredContent;
    }

    resetTime() {
        this.time = Math.trunc(new Date().getTime()/1000);
    }
}
registerType(ChatEntry);
